### ISSUE TITLE
code/chapter7/patterns/pool/pool.go does not restrict resources

### ISSUE THREAD

**Issue by greggkp** [goinaction/code#25]
If I've understood the code correctly, line 54: p.factory() is not limited to the number of resources available (2). Suggest moving resource allocation to New() and removing the select in Acquire() so that the channel blocks.

**Comment by ghost** [comment]
I came across this too. It does indeed appear that Acquire() allows an arbitrary number of resources to be created when the channel is emptied. What it should do, though, is block until a resource is available again.

I attempted a workaround that will allow Acquire to block when the number of resources reaches the channel's capacity, as well as continue to only allocate resources only when needed:

```
func (p *Pool) Acquire() (io.Closer, error) {
	select {
	case resource, ok := <-p.resources:
		return onResourceAvailable(resource, ok)

	default:
		p.createLock.Lock()
		defer p.createLock.Unlock()
		if p.createdCount >= cap(p.resources) {
			//capacity reached, wait for resource to become available
			resource, ok := <-p.resources
			return onResourceAvailable(resource, ok)
		}

		//capacity not yet reached, so create new resource
		log.Println("Acquire:", "new")
		p.createdCount++
		return p.factory()
	}
}

func onResourceAvailable(resource io.Closer, ok bool) (io.Closer, error) {
	log.Println("Acquire:", "existing")
	if !ok {
		return nil, ErrPoolClosed
	}
	return resource, nil
}
```

You will, of course, need to add `createLock sync.Mutex` and `createdCount int` to your Pool type definition.

There may be an idiomatic way of solving this problem, for now this seems to work for me.

**Comment by greggkp** [comment]
Thanks, that looks like it should work, and it keeps with the original code's intent of lazily allocating the resources.

I had a go at that approach too and decided that lazy resource allocation is complicated and it felt odd that I needed a mutex to read from a channel... So, failing to find a way to avoid that, I moved the resource allocation to the New() function and just blocked inside Acquire(). That did the trick, but it is no longer allocating resources lazily!

**Comment by ghost** [comment]
Indeed. It is not ideal to be holding on to the mutex while a resource is not available, but we do need it when we're checking capacity (and the subsequent create + count increment, if createdCount < cap).

It may help to unlock the mutex right after checking capacity *if* we're going to wait on the channel anyways. I'll tinker a little more and see whether I can improve it. Will post if I discover anything useful.

**Comment by ghost** [comment]
Tinkered a bit more. Best I came up with is this:

```
func (p *Pool) Acquire() (io.Closer, error) {
	select {
	case resource, ok := <-p.resources:
		return onResourceAvailable(resource, ok)

	default:
		//standard double-checked locking
		if p.createdCount < cap(p.resources) {
			p.createLock.Lock()
			if p.createdCount < cap(p.resources) {
				//capacity not yet reached, so create new resource
				defer p.createLock.Unlock()
				log.Println("Acquire:", "new")
				p.createdCount++
				return p.factory()
			}
			p.createLock.Unlock()
		}
		//capacity reached, wait for resource to become available
		resource, ok := <-p.resources
		return onResourceAvailable(resource, ok)
	}
}
```

We can use standard double-checked locking to check whether we've reached capacity. This way, we can block on the channel receive without the overhead of holding the lock.

There will naturally be some initial calls to Acquire where goroutines block unnecessarily while trying to acquire the lock, but will quickly abandon it once they see that the created count matches the capacity, and they will then resort to block on the channel receive.

Any goroutines thereafter that check whether capacity has been reached (and it has been reached) will not acquire the lock.

You could even get rid of all the code outside the default case as it is now redundant. Might be nice to benchmark it and see whether there is a measurable effect on performance by leaving it in.

One thing to note though, is that depending on how the golang compiler inlines and initialises variables, this may still have race conditions (as I discovered with `go build --race`). The safer approach is to block as I did in my previous solution, perhaps with the added optimisation of unlocking the mutex if we've reached capacity and we'll block on the channel receive, like so:

```
p.createLock.Lock()
if p.createdCount >= cap(p.resources) {
    p.createLock.Unlock()
    //capacity reached, wait for resource to become available
    resource, ok := <-p.resources
    return onResourceAvailable(resource, ok)
}
defer p.createLock.Unlock()

//capacity not yet reached, so create new resource
p.createdCount++
return p.factory()
```

**Comment by daliny** [comment]
line 29: `size uint`，line 30: `size <= 0`, the **uint** number always greater  than 0 ...
`func New(fn func() (io.Closer, error), **size uint**) (*Pool, error) {
	if **size <= 0** {
		return nil, errors.New("Size value too small.")
	}

	return &Pool{
		factory:   fn,
		resources: make(chan io.Closer, size),
	}, nil
}`

**Comment by rammanis97** [comment]
Thanks ,for this.
